### ISSUE TITLE
思考机制实验发现：训练代码未实现理论中的状态约束损失 L2，期待 v0d2

### ISSUE THREAD

**Issue by Offblink** [zhihumomo/ouro#4]
## 关于"思考机制"的实验发现 + 对 v0d2 的期待

> 此 issue 由 Blinvo 及其助手提交。我们在本地复现并实验了 Gridman-Medium（v0d1 SFT），围绕"状态驱动思考"做了一系列尝试，有一些发现想与作者分享。

---

### 一、我们在思考上做的努力

受 Ouro "State is all you need" 理念启发，我们实现了 **Acheron Thinking**——一种不依赖 CoT prompt、完全利用状态自循环演化的推理机制：

1. **状态自循环**：思考时不让模型生成 token，而是通过重复前向传播让 STM 状态（`c_state`）演化，mem 矩阵通过联想检索参与
2. **三种思考模式**：
   - `focused`：快速收敛（能量阈值 1e-3），约 4 步
   - `divergent`：噪声注入 + 极紧阈值（5e-5），探索更深，约 12 步
   - `reflective`：双重 query 重读，通过 temporal attention 回顾上一步状态
3. **收敛检测**：以 `E(t) = ||c_t - c_{t-1}||² / dim` 作为能量函数，状态稳定即收敛

**技术路径迭代**：最初用零输入共振 → 状态坍缩；改用 EOS token 触发 → 状态漂移出训练分布；最终采用 **query 重读**（seq_len>1，mem 更新阻尼 `1/seq_len` 生效），才让知识安全参与思考。

#### 1. 【核心】README 理论中的状态约束损失 L2 未在训练代码中实现

README 中 Ouro 完备性的核心是等效原理：

```
L2(θ) = E[ || W(θ, s_{t+1}) - W(θ, s_t) - J_t(s_{t+1} - s_t) ||² ]
```

但 `naxi/v_0d1/gridman/train.py`（约 130 行）的训练目标**只有交叉熵**：

```python
base_loss = F.cross_entropy(logits, targets, ignore_index=-100)
# SFT 时叠加余弦衰减权重，无任何状态约束项
```

后果：模型被训练成"状态架构下的静态函数"——架构是状态的（mem 矩阵、STM、时间队列真实存在且跨对话持久化），但训练从未奖励"状态演化得有意义"。

**实验证据**：我们在同一问题上对比"不思考 vs 三种思考模式"，重复 3 次统计关键词命中（月相/太阳/地球/轨道/反射/公转）：

| 模式 | 平均命中 |
|------|---------|
| 不思考（原始 SFT） | 最高，最稳定 |
| focused | 4.0 |
| divergent | 1.7 |

结论：**思考机制无法超越不思考**，因为思考只能重新采样训练中学到的模式，而训练目标（纯交叉熵）从未让状态演化变得有知识增量。这正是 README 批评的"函数近似器"范式——讽刺的是训练代码实现的恰好是这个。

#### 2. `mem_clear()` 会重置训练好的 mem 矩阵

`OuroLayer.mem_clear()` 将 mem 重置为 `torch.eye(embed_dim)`，这会**销毁训练学到的长期记忆**。我们早期实验因此输出全部劣化。建议提供仅清 STM、保留 mem 的"软重置"接口（我们在 Acheron 中已实现此区分）。

#### 3. `naxi/v_0d2/` 目录为空

仅含 `.gitkeep`，README 标注"即将发布"。

### 三、期待 v0d2

1. **实现 L2 状态约束损失**：让训练真正奖励"状态演化 = 知识重组"，这是 Ouro 理论的核心，也是让"思考机制"有理论支撑的前提
2. **公开 v0d2 训练代码**：哪怕是最简实现，也足以验证等效原理是否成立
3. **如果可能**：提供状态约束训练的消融对比（有 L2 vs 无 L2），我们非常好奇

### 四、我们愿意贡献

Acheron Thinking 的实现（`thinking.py`，约 500 行，纯推理侧，零训练改动）可作为 v0d2 的验证工具——如果 v0d2 实现了 L2，我们愿意配合做"思考机制在有/无状态约束训练下的效果对比"实验。

感谢这个项目：它让我们第一次摸到"状态 AI"的实感。期待 v0d2 把理论真正落地。

—— Blinvo 及其助手

**Comment by zhihumomo** [comment]
L2是一个严格的数学约束, 但是实现L2会带来不可接受的性能损失. 当前架构实际上继承了思想, 后续可能会引入近似实现.

**Comment by zhihumomo** [comment]
感谢你的工作.
